File: storage/src/main/java/pl/expensive/storage/Wallet.java
package pl.expensive.storage;

import com.google.auto.value.AutoValue;

import java.util.UUID;

@AutoValue
public abstract class Wallet {
    public static Wallet create(UUID uuid, String name, Currency currency) {
        return new AutoValue_Wallet(uuid, name, currency);
    }

    public abstract UUID uuid();
    public abstract String name();
    public abstract Currency currency();
}

File: storage/src/main/java/pl/expensive/storage/Wallet.kt
package pl.expensive.storage

import android.os.Parcel
import android.os.Parcelable
import java.util.*

data class Wallet(val uuid: UUID,
                  val name: String,
                  val currency: Currency) : Parcelable {

    constructor(parcelIn: Parcel) : this(
            parcelIn.readString().toUUID(),
            parcelIn.readString(),
            parcelIn.readParcelable(Currency.CREATOR))

    override fun writeToParcel(dest: Parcel, flags: Int) {
        with(dest) {
            writeString(uuid.toString())
            writeString(name)
            writeParcelable(currency, 0)
        }
    }

    override fun describeContents() = 0

    companion object {
        @JvmField @Suppress("unused")
        val CREATOR = createParcel(::Wallet)
    }
}
